File: experiment-15-sonnet5/brazil/runs/language=typescript_model=opus-4.8_prompt=bdd/rep1/src/queries/teams.ts
/**
 * Team-centric queries: win/draw/loss records, goals for/against, and
 * home vs away splits, computed from played matches.
 */
import { teamMatches } from "../normalize.js";
import type { Match, TeamRecord } from "../types.js";
import { filterMatches, type MatchFilter } from "./filters.js";

export interface TeamStats extends TeamRecord {
  goalDifference: number;
  winRate: number;
  home: TeamRecord;
  away: TeamRecord;
  competitions: string[];
}

function emptyRecord(team: string): TeamRecord {
  return {
    team,
    played: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goalsFor: 0,
    goalsAgainst: 0,
    points: 0,
  };
}

function applyResult(record: TeamRecord, scored: number, conceded: number): void {
  record.played += 1;
  record.goalsFor += scored;
  record.goalsAgainst += conceded;
  if (scored > conceded) {
    record.wins += 1;
    record.points += 3;
  } else if (scored === conceded) {
    record.draws += 1;
    record.points += 1;
  } else {
    record.losses += 1;
  }
}

/**
 * Aggregate a team's performance over the matches selected by `filter`.
 * `filter.team` is set from `team`, so callers typically pass season and/or
 * competition constraints in `filter`.
 */
export function teamStats(all: Match[], team: string, filter: MatchFilter = {}): TeamStats {
  const matches = filterMatches(all, { ...filter, team, playedOnly: true });

  const total = emptyRecord(team);
  const home = emptyRecord(team);
  const away = emptyRecord(team);
  const competitions = new Set<string>();

  for (const m of matches) {
    competitions.add(m.competition);
    const homeGoals = m.homeGoals as number;
    const awayGoals = m.awayGoals as number;
    if (teamMatches(m.homeTeam, team)) {
      applyResult(total, homeGoals, awayGoals);
      applyResult(home, homeGoals, awayGoals);
    } else {
      applyResult(total, awayGoals, homeGoals);
      applyResult(away, awayGoals, homeGoals);
    }
  }

  return {
    ...total,
    goalDifference: total.goalsFor - total.goalsAgainst,
    winRate: total.played > 0 ? total.wins / total.played : 0,
    home,
    away,
    competitions: [...competitions].sort(),
  };
}

File: experiments/adrianco/experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=TDD/rep1/lib/brazilian_soccer/queries/teams.ex
defmodule BrazilianSoccer.Queries.Teams do
  @moduledoc """
  Team-centric queries: win/loss/draw records, goals for/against and win rate,
  optionally restricted by season, competition or venue (home/away).
  """

  alias BrazilianSoccer.{Dataset, Match, TeamName}
  alias BrazilianSoccer.Queries.{Matches, Source}

  @doc """
  Compute a team's record.

  Options: `:season`, `:competition`, and `:venue` (`:home`, `:away`, `:all`).
  Returns a map with played/wins/draws/losses, goals_for/against,
  goal_difference, points and win_rate.
  """
  @spec record(Dataset.t(), binary(), keyword()) :: map()
  def record(%Dataset{} = ds, team, opts \\ []) do
    key = TeamName.base(team)
    venue = Keyword.get(opts, :venue, :all)

    find_opts =
      opts
      |> Keyword.take([:season, :competition])
      |> Keyword.put(:team, team)

    matches =
      ds
      |> Matches.find(find_opts)
      |> Source.primary_per_season()
      |> Enum.filter(&venue_match?(&1, key, venue))
      |> Enum.filter(&(Match.winner(&1) != nil))

    Enum.reduce(matches, base(team), fn m, acc -> tally(m, key, acc) end)
    |> finalize()
  end

  @doc "Compare two teams: their individual records plus the head-to-head."
  @spec compare(Dataset.t(), binary(), binary()) :: map()
  def compare(%Dataset{} = ds, team_a, team_b) do
    %{
      team_a: record(ds, team_a),
      team_b: record(ds, team_b),
      head_to_head: Matches.head_to_head(ds, team_a, team_b)
    }
  end

  defp base(team) do
    %{
      team: TeamName.clean(team),
      played: 0,
      wins: 0,
      draws: 0,
      losses: 0,
      goals_for: 0,
      goals_against: 0
    }
  end

  defp venue_match?(%Match{home_base: hk}, key, :home), do: hk == key
  defp venue_match?(%Match{away_base: ak}, key, :away), do: ak == key
  defp venue_match?(_m, _key, _), do: true

  defp tally(%Match{} = m, key, acc) do
    {gf, ga} =
      if m.home_base == key do
        {m.home_goals, m.away_goals}
      else
        {m.away_goals, m.home_goals}
      end

    result =
      cond do
        gf > ga -> :wins
        ga > gf -> :losses
        true -> :draws
      end

    acc
    |> Map.update!(:played, &(&1 + 1))
    |> Map.update!(:goals_for, &(&1 + gf))
    |> Map.update!(:goals_against, &(&1 + ga))
    |> Map.update!(result, &(&1 + 1))
  end

  defp finalize(acc) do
    acc
    |> Map.put(:goal_difference, acc.goals_for - acc.goals_against)
    |> Map.put(:points, acc.wins * 3 + acc.draws)
    |> Map.put(:win_rate, win_rate(acc.wins, acc.played))
  end

  defp win_rate(_wins, 0), do: 0.0
  defp win_rate(wins, played), do: wins / played
end
